src/bls12_381/ec.rs

```diff
@@ -130,18 +158,56 @@ macro_rules! curve_impl {
                     let mut y2 = self.y;
                     y2.square();
 
-                    let mut x3b = self.x;
-                    x3b.square();
-                    x3b.mul_assign(&self.x);
-                    x3b.add_assign(&Self::get_coeff_b());
-
-                    y2 == x3b
+                    y2 == y2_from_x(self.x)
                 }
             }
 
             fn is_in_correct_subgroup_assuming_on_curve(&self) -> bool {
                 self.mul($scalarfield::char()).is_zero()
             }
+
+            fn sw_encode(t: $basefield) -> Self {
+                // handle the case t == 0
+                if t.is_zero() { return Self::zero() };
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&get_coeff_b());
+                w.add_assign(&$basefield::one());
+                // handle the case t^2 + b + 1 == 0
+                if w.is_zero() { return Self::one() };
```

Comment:
In my proposal, I wanted to send the "positive" t value to Self::one, and the "negative" t value to negative Self::one. This might seem silly, but it maintains the relationship that `sw_encode(a) = -sw_encode(-a)`, so it's elegant and less arbitrary.